File: archestra-bench/tasks.py
# /// script
# requires-python = ">=3.11"
# dependencies = []
# ///
"""Self-contained benchmark task model + loader.

A task lives in its own directory:

    tasks/<id>/task.toml     declarative definition (stages, result schema, verifier, artifact)
    tasks/<id>/verifier.py   pytest verifier, run out of band against the submission
    tasks/<id>/inputs/       files staged into the agent's sandbox (also readable by the verifier)
    tasks/<id>/expected/     verifier-only ground truth; NEVER staged to the agent

A task is an ordered list of conversation **stages** (a "user asks X" turn, then optional "user
corrects to Y" turns). The agent solves it with whatever tools/skills its environment provides and
hands in its answer by calling the benchmark MCP's `submit_result` tool -- so a task declares the
JSON-schema that answer must match (`result_schema`).

A task whose deliverable is a *file* (not a JSON value) sets `artifact_key`: the result property
naming the file the agent exported via `download_file`. The harness downloads that artifact and
hands its bytes to the verifier as `BENCH_OUTPUT` (see run.py / verify.py).

The verifier assets are never staged anywhere the agent can reach. Staged files are confined to
`inputs/` at load time, so a precomputed answer in `expected/` can never leak to the agent.
"""

from __future__ import annotations

import re
from collections.abc import Mapping
from dataclasses import dataclass, field
from pathlib import Path, PurePosixPath
from typing import Any
from urllib.parse import unquote, urlsplit

import tomlconf

_FILE_PLACEHOLDER = re.compile(r"\{\{file:([^}]+)\}\}")


@dataclass(frozen=True)
class StagedFile:
    """an input file the agent is allowed to see, delivered as a chat file-part on its stage.

    `src` is a relative path under the task's `inputs/` dir; `dest` is the absolute sandbox path it
    auto-stages to (the stage message references it)."""

    src: str
    dest: str
    mime_type: str = "application/octet-stream"


@dataclass(frozen=True)
class Stage:
    """one user turn: `text` is the message, `files` are delivered with it as chat file-parts.

    Later stages model the user changing or refining their ask within the same conversation."""

    text: str
    files: tuple[StagedFile, ...] = ()


@dataclass(frozen=True)
class Verifier:
    """how the harness verifies the submission, OUT of band (never staged where the agent can reach).

    The verifier runs as a pytest file in an ephemeral uv env. It reads the submission from
    `BENCH_RESULT`, optional fixtures from `BENCH_FIXTURES` (a dir with `inputs/` and `expected/`
    subdirectories), and an optional downloaded artifact from `BENCH_OUTPUT` (see verify.py)."""

    deps: tuple[str, ...] = ()  # pip-style requirements for the ephemeral uv env
    test_file: str = "verifier.py"  # pytest file, relative to the task dir
    env: dict[str, str] = field(default_factory=dict)  # extra env (time limits, etc.)


@dataclass(frozen=True)
class Task:
    """declarative description of one benchmark task, loaded from `tasks/<id>/task.toml`."""

    id: str
    dir: Path  # absolute path to the task's directory
    stages: tuple[Stage, ...]
    result_schema: dict[str, Any]  # JSON-schema the submitted result must match
    verifier: Verifier
    artifact_key: str | None = None  # result property naming the produced artifact filename
    max_format_attempts: int = 3  # submit_result self-correction budget
    state_rest: tuple[str, ...] = ()  # backend REST GET paths to snapshot into BENCH_STATE

    @property
    def inputs_dir(self) -> Path:
        return self.dir / "inputs"

    @property
    def expected_dir(self) -> Path:
        return self.dir / "expected"


_DEFAULT_MAX_FORMAT_ATTEMPTS = 3


def load_task(task_dir: Path) -> Task:
    """Parse `task_dir/task.toml` into a Task; validate loudly (id, schema, staged-file confinement)."""
    task_id = task_dir.name
    ctx = f"task {task_id!r}"
    if not tomlconf.is_slug(task_id):
        raise SystemExit(f"{ctx}: task dir name must be lowercase alphanumeric with dashes (slug-safe)")
    toml_path = task_dir / "task.toml"
    if not toml_path.is_file():
        raise SystemExit(f"{ctx}: missing {toml_path}")
    data = tomlconf.parse_toml(toml_path)

    stage_rows = tomlconf.rows(data, "stages", ctx)
    if not stage_rows:
        raise SystemExit(f"{ctx}: task declares no stages")
    stages = tuple(_stage(row, ctx, task_dir) for row in stage_rows)

    schema = tomlconf.table(data, "result_schema", ctx)
    if not isinstance(schema, dict):
        raise SystemExit(f"{ctx}: result_schema must be a table")

    max_attempts = tomlconf.req_int(data, "max_format_attempts", ctx, default=_DEFAULT_MAX_FORMAT_ATTEMPTS)
    if max_attempts < 1:
        raise SystemExit(f"{ctx}: max_format_attempts must be >= 1, got {max_attempts}")

    verifier = _verifier(tomlconf.table(data, "verifier", ctx, default={}), f"{ctx} [verifier]", task_dir)
    state = tomlconf.table(data, "state", ctx, default={})
    state_rest = tuple(_state_path(p, f"{ctx} [state]") for p in tomlconf.strs(state, "rest", f"{ctx} [state]"))

    return Task(
        id=task_id,
        dir=task_dir,
        stages=stages,
        result_schema=dict(schema),
        verifier=verifier,
        artifact_key=tomlconf.opt_str(data, "artifact_key", ctx),
        max_format_attempts=max_attempts,
        state_rest=state_rest,
    )


def _state_path(path: str, ctx: str) -> str:
    """A state-capture entry must be a relative backend API path the harness GETs after the run.

    Reject anything that could escape the backend or the API surface: an absolute URL (scheme/host),
    a path outside `/api/`, or a `..` traversal. Query strings and the runtime `{{cell}}`/`{{agent_id}}`
    placeholders are allowed -- they're substituted at capture time (see run.py)."""
    parts = urlsplit(path)
    if parts.scheme or parts.netloc:
        raise SystemExit(f"{ctx}: rest path {path!r} must be a relative /api/ path, not an absolute URL")
    if not parts.path.startswith("/api/"):
        raise SystemExit(f"{ctx}: rest path {path!r} must start with /api/")
    if ".." in unquote(parts.path).split("/"):  # unquote first: reject `/api/../x` and `/api/%2e%2e/x`
        raise SystemExit(f"{ctx}: rest path {path!r} must not contain a '..' segment")
    return path


def _stage(row: Mapping[str, Any], ctx: str, task_dir: Path) -> Stage:
    text = _expand_files(tomlconf.req_str(row, "text", ctx), task_dir, ctx)
    files = tuple(_staged_file(f, ctx, task_dir / "inputs") for f in tomlconf.rows(row, "files", ctx))
    return Stage(text=text, files=files)


def _expand_files(text: str, task_dir: Path, ctx: str) -> str:
    """Expand `{{file:<relpath>}}` placeholders with the referenced file's text content.

    Used to inline a fixture (e.g. a CSV) into the prompt for providers/tasks that cannot stage a
    file into the sandbox. The path is confined to the task dir so a task can never inline an
    out-of-tree file."""
    base = task_dir.resolve()

    def repl(match: re.Match[str]) -> str:
        rel = match.group(1).strip()
        target = (task_dir / rel).resolve()
        if target != base and base not in target.parents:
            raise SystemExit(f"{ctx}: file placeholder {rel!r} escapes the task dir")
        if not target.is_file():
            raise SystemExit(f"{ctx}: file placeholder {rel!r} does not exist")
        return target.read_text(encoding="utf-8")

    return _FILE_PLACEHOLDER.sub(repl, text)


def _staged_file(row: Mapping[str, Any], ctx: str, inputs_dir: Path) -> StagedFile:
    src = tomlconf.req_str(row, "src", ctx)
    _check_under_inputs(src, inputs_dir, ctx)
    return StagedFile(
        src=src,
        dest=tomlconf.req_str(row, "dest", ctx),
        mime_type=tomlconf.req_str(row, "mime_type", ctx, default="application/octet-stream"),
    )


def _check_under_inputs(src: str, inputs_dir: Path, ctx: str) -> None:
    """A staged source must resolve to an existing file strictly under `inputs/` -- never an
    absolute path or a `..` escape into the verifier-only `expected/` dir or out of the task tree."""
    if PurePosixPath(src).is_absolute() or src.startswith("/"):
        raise SystemExit(f"{ctx}: staged file src {src!r} must be relative (under inputs/)")
    base = inputs_dir.resolve()
    target = (inputs_dir / src).resolve()
    if target != base and base not in target.parents:
        raise SystemExit(f"{ctx}: staged file src {src!r} escapes inputs/")
    if not target.is_file():
        raise SystemExit(f"{ctx}: staged file {inputs_dir / src} does not exist")


def _verifier(tbl: Mapping[str, Any], ctx: str, task_dir: Path) -> Verifier:
    defaults = Verifier()
    test_file = tomlconf.req_str(tbl, "test_file", ctx, default=defaults.test_file)
    if PurePosixPath(test_file).is_absolute() or test_file.startswith("/"):
        raise SystemExit(f"{ctx}: test_file {test_file!r} must be relative (under the task dir)")
    base = task_dir.resolve()
    target = (task_dir / test_file).resolve()
    if target != base and base not in target.parents:
        raise SystemExit(f"{ctx}: test_file {test_file!r} escapes the task dir")
    if not target.is_file():
        raise SystemExit(f"{ctx}: verifier {task_dir / test_file} does not exist")
    return Verifier(
        deps=tuple(tomlconf.strs(tbl, "deps", ctx)),
        test_file=test_file,
        env=tomlconf.str_map(tbl, "env", ctx),
    )

File: archestra-bench/runner/src/config/tasks.rs
use std::path::{Path, PathBuf};

use super::toml_util::{self, TomlTable};
use super::types::{Stage, StagedFile, Task, Verifier};

const FILE_PLACEHOLDER_RE: &str = r"\{\{file:([^}]+)\}\}";
const DEFAULT_MAX_FORMAT_ATTEMPTS: i64 = 3;

#[derive(Debug, thiserror::Error)]
#[error("task config error: {0}")]
pub struct TaskConfigError(pub String);

impl From<toml_util::TomlError> for TaskConfigError {
    fn from(e: toml_util::TomlError) -> Self {
        Self(e.to_string())
    }
}

pub fn load_task(task_dir: &Path) -> Result<Task, TaskConfigError> {
    let task_id = task_dir
        .file_name()
        .ok_or_else(|| TaskConfigError("task dir has no name".to_string()))?
        .to_string_lossy()
        .to_string();
    let ctx = format!("task {task_id:?}");

    if !archestra_bench_core::is_slug(&task_id) {
        return Err(TaskConfigError(format!(
            "{ctx}: task dir name must be lowercase alphanumeric with dashes (slug-safe)"
        )));
    }

    let toml_path = task_dir.join("task.toml");
    if !toml_path.is_file() {
        return Err(TaskConfigError(format!("{ctx}: missing {toml_path:?}")));
    }

    let data = toml_util::parse_toml_file(&toml_path)?;

    let stage_rows = toml_util::rows(&data, "stages", &ctx)?;
    if stage_rows.is_empty() {
        return Err(TaskConfigError(format!("{ctx}: task declares no stages")));
    }
    let mut stages = Vec::with_capacity(stage_rows.len());
    for (i, row) in stage_rows.iter().enumerate() {
        let row_ctx = format!("{ctx} [[stages]][{i}]");
        stages.push(load_stage(row, &row_ctx, task_dir)?);
    }

    let schema = toml_util::table(&data, "result_schema", &ctx)?;
    let schema_value = toml_table_to_json_value(&schema);

    let max_attempts = toml_util::req_int(
        &data,
        "max_format_attempts",
        &ctx,
        DEFAULT_MAX_FORMAT_ATTEMPTS,
    )? as usize;
    if max_attempts < 1 {
        return Err(TaskConfigError(format!(
            "{ctx}: max_format_attempts must be >= 1, got {max_attempts}"
        )));
    }

    let verifier_table = toml_util::table_with_default(&data, "verifier", &ctx)?;
    let verifier = load_verifier(&verifier_table, &format!("{ctx} [verifier]"), task_dir)?;

    let state_table = toml_util::table_with_default(&data, "state", &ctx)?;
    let state_rest = load_state_rest(&state_table, &format!("{ctx} [state]"))?;

    Ok(Task {
        id: task_id,
        dir: task_dir.to_path_buf(),
        stages,
        result_schema: schema_value,
        verifier,
        artifact_key: toml_util::opt_str(&data, "artifact_key", &ctx)?,
        max_format_attempts: max_attempts,
        state_rest,
    })
}

fn load_stage(row: &TomlTable, ctx: &str, task_dir: &Path) -> Result<Stage, TaskConfigError> {
    let text = toml_util::req_str(row, "text", ctx)?;
    let text = expand_files(&text, task_dir, ctx)?;
    let files = load_staged_files(row, ctx, &task_dir.join("inputs"))?;
    Ok(Stage { text, files })
}

fn expand_files(text: &str, task_dir: &Path, ctx: &str) -> Result<String, TaskConfigError> {
    let base = task_dir;
    let re = regex::Regex::new(FILE_PLACEHOLDER_RE).expect("valid regex");
    let mut errors = Vec::new();
    let result = re.replace_all(text, |caps: &regex::Captures| {
        let rel = caps[1].trim();
        let target = match resolve_under(base, rel) {
            Ok(t) => t,
            Err(e) => {
                errors.push(e);
                return "".to_string();
            }
        };
        if !target.is_file() {
            errors.push(format!("{ctx}: file placeholder {rel:?} does not exist"));
            return "".to_string();
        }
        match std::fs::read_to_string(&target) {
            Ok(s) => s,
            Err(e) => {
                errors.push(format!("{ctx}: cannot read file placeholder {rel:?}: {e}"));
                "".to_string()
            }
        }
    });
    if !errors.is_empty() {
        return Err(TaskConfigError(errors.into_iter().next().unwrap()));
    }
    Ok(result.into_owned())
}

fn load_staged_files(
    row: &TomlTable,
    ctx: &str,
    inputs_dir: &Path,
) -> Result<Vec<StagedFile>, TaskConfigError> {
    let rows = toml_util::rows(row, "files", ctx)?;
    let mut files = Vec::with_capacity(rows.len());
    for (i, file_row) in rows.iter().enumerate() {
        let file_ctx = format!("{ctx} [[files]][{i}]");
        let src = toml_util::req_str(file_row, "src", &file_ctx)?;
        check_under_inputs(&src, inputs_dir, &file_ctx)?;
        files.push(StagedFile {
            src,
            dest: toml_util::req_str(file_row, "dest", &file_ctx)?,
            mime_type: toml_util::req_str_with_default(
                file_row,
                "mime_type",
                &file_ctx,
                "application/octet-stream",
            )?,
        });
    }
    Ok(files)
}

fn check_under_inputs(src: &str, inputs_dir: &Path, ctx: &str) -> Result<(), TaskConfigError> {
    if src.starts_with('/') {
        return Err(TaskConfigError(format!(
            "{ctx}: staged file src {src:?} must be relative (under inputs/)"
        )));
    }
    let target = resolve_under(inputs_dir, src).map_err(TaskConfigError)?;
    if !target.is_file() {
        return Err(TaskConfigError(format!(
            "{ctx}: staged file {target:?} does not exist"
        )));
    }
    Ok(())
}

fn resolve_under(base: &Path, rel: &str) -> Result<PathBuf, String> {
    let target = base.join(rel);
    let target = target
        .canonicalize()
        .map_err(|e| format!("cannot resolve {rel:?}: {e}"))?;
    let base_canon = base
        .canonicalize()
        .map_err(|e| format!("cannot resolve base dir: {e}"))?;
    if target != base_canon && !target.starts_with(&base_canon) {
        return Err(format!("path {rel:?} escapes base directory"));
    }
    Ok(target)
}

fn load_verifier(tbl: &TomlTable, ctx: &str, task_dir: &Path) -> Result<Verifier, TaskConfigError> {
    let test_file = toml_util::req_str_with_default(tbl, "test_file", ctx, "verifier.py")?;
    if test_file.starts_with('/') {
        return Err(TaskConfigError(format!(
            "{ctx}: test_file {test_file:?} must be relative (under the task dir)"
        )));
    }
    let target = resolve_under(task_dir, &test_file).map_err(|e| {
        TaskConfigError(format!(
            "{ctx}: test_file {test_file:?} escapes the task dir: {e}"
        ))
    })?;
    if !target.is_file() {
        return Err(TaskConfigError(format!(
            "{ctx}: verifier {target:?} does not exist"
        )));
    }
    Ok(Verifier {
        deps: toml_util::strs(tbl, "deps", ctx)?,
        test_file,
        env: toml_util::str_map(tbl, "env", ctx)?,
    })
}

fn load_state_rest(tbl: &TomlTable, ctx: &str) -> Result<Vec<String>, TaskConfigError> {
    let paths = toml_util::strs(tbl, "rest", ctx)?;
    for p in &paths {
        validate_state_path(p, ctx)?;
    }
    Ok(paths)
}

fn validate_state_path(path: &str, ctx: &str) -> Result<(), TaskConfigError> {
    // Validate the *path* component (query strings and the {{cell}}/{{agent_id}} placeholders are
    // substituted later and allowed), and decode percent-escapes before the `..` check so
    // `/api/%2e%2e/x` is rejected just like `/api/../x`.
    let path_part = path.split('?').next().unwrap_or(path);
    if path_part.contains("://") || path_part.starts_with("//") {
        return Err(TaskConfigError(format!(
            "{ctx}: rest path {path:?} must be a relative /api/ path, not an absolute URL"
        )));
    }
    if !path_part.starts_with("/api/") {
        return Err(TaskConfigError(format!(
            "{ctx}: rest path {path:?} must start with /api/"
        )));
    }
    let decoded = percent_encoding::percent_decode_str(path_part)
        .decode_utf8_lossy()
        .into_owned();
    if decoded.split('/').any(|seg| seg == "..") {
        return Err(TaskConfigError(format!(
            "{ctx}: rest path {path:?} must not contain a '..' segment"
        )));
    }
    Ok(())
}

fn toml_table_to_json_value(table: &TomlTable) -> serde_json::Value {
    let mut map = serde_json::Map::new();
    for (k, v) in table {
        map.insert(k.clone(), toml_value_to_json_value(v));
    }
    serde_json::Value::Object(map)
}

fn toml_value_to_json_value(value: &toml::Value) -> serde_json::Value {
    match value {
        toml::Value::String(s) => serde_json::Value::String(s.clone()),
        toml::Value::Integer(i) => serde_json::Value::Number((*i).into()),
        toml::Value::Float(f) => serde_json::Number::from_f64(*f)
            .map_or(serde_json::Value::Null, serde_json::Value::Number),
        toml::Value::Boolean(b) => serde_json::Value::Bool(*b),
        toml::Value::Datetime(d) => serde_json::Value::String(d.to_string()),
        toml::Value::Array(arr) => {
            serde_json::Value::Array(arr.iter().map(toml_value_to_json_value).collect())
        }
        toml::Value::Table(t) => toml_table_to_json_value(t),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn state_path_accepts_api_path_with_query_and_placeholders() {
        validate_state_path("/api/skills?search=prime-counter-{{cell}}&limit=100", "ctx").unwrap();
        validate_state_path("/api/agents/{{agent_id}}/tools", "ctx").unwrap();
    }

    #[test]
    fn state_path_rejects_plain_and_encoded_traversal() {
        assert!(validate_state_path("/api/../x", "ctx").is_err());
        assert!(validate_state_path("/api/%2e%2e/x", "ctx").is_err());
        assert!(validate_state_path("/api/%2E%2E/x", "ctx").is_err());
    }

    #[test]
    fn state_path_rejects_non_api_and_absolute_url() {
        assert!(validate_state_path("/other", "ctx").is_err());
        assert!(validate_state_path("http://host/api/x", "ctx").is_err());
        assert!(validate_state_path("//host/api/x", "ctx").is_err());
    }
}
